File: AlgorithmsPackPythonSolutions/Min Cost to Connect All Points.py
class UnionFind:
    def __init__(self, n):
        self.forests = n
        self.rank = [1] * n
        self.parent = list(range(n))

    def find_set(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find_set(self.parent[x])
        return self.parent[x]

    def union_set(self, x, y):
        x = self.find_set(x)
        y = self.find_set(y)
        if x != y:
            self.link(x, y)
            self.forests -= 1
        return x != y

    def link(self, x, y):
        if self.rank[x] > self.rank[y]:
            x, y = y, x
        self.parent[x] = y
        if self.rank[x] == self.rank[y]:
            self.rank[y] += 1


class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        edgeList = []
        for i in range(n):
            for j in range(i, n):
                cost = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1])
                edgeList.append((i, j, cost))

        edgeList.sort(key=lambda e: e[2])
        uf = UnionFind(n)
        mstCost = 0

        for e in edgeList:
            if uf.union_set(e[0], e[1]):
                mstCost += e[2]

        return mstCost

File: AlgorithmsPackJavaSolutions/Min Cost to Connect All Points.java
import java.util.*;

class Solution {
    private class UnionFind {
        private int forests;
        private int[] rank;
        private int[] parent;

        public UnionFind(int n) {
            forests = n;
            rank = new int[n];
            parent = new int[n];
            for (int i = 0; i < n; i++) {
                rank[i] = 1;
                parent[i] = i;
            }
        }

        public int findSet(int x) {
            if (parent[x] == x)
                return x;
            return parent[x] = findSet(parent[x]);
        }

        public boolean unionSet(int x, int y) {
            x = findSet(x);
            y = findSet(y);
            if (x != y) {
                link(x, y);
                forests--;
            }
            return x != y;
        }

        private void link(int x, int y) {
            if (rank[x] > rank[y])
                parent[y] = x;
            else {
                parent[x] = y;
                if (rank[x] == rank[y])
                    rank[y]++;
            }
        }
    }

    private class Edge implements Comparable<Edge> {
        private int from;
        private int to;
        private int weight;

        public Edge(int from, int to, int weight) {
            this.from = from;
            this.to = to;
            this.weight = weight;
        }

        public int compareTo(Edge e) {
            return Integer.compare(weight, e.weight);
        }
    }

    public int minCostConnectPoints(int[][] points) {
        int n = points.length;
        List<Edge> edgeList = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int cost = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1]);
                edgeList.add(new Edge(i, j, cost));
            }
        }

        return kruskal(edgeList, n);
    }

    private int kruskal(List<Edge> edgeList, int n) {
        Collections.sort(edgeList);
        UnionFind uf = new UnionFind(n);
        int mstCost = 0;

        for (Edge e : edgeList) {
            if (uf.unionSet(e.from, e.to))
                mstCost += e.weight;
        }

        return mstCost;
    }
}
